File: app/src/main/java/de/hdc/kspchecklist/DetailTouchHelper.java
package de.hdc.kspchecklist;

import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;

/**
 * Created by DerTroglodyt on 2017-02-18 10:12.
 * Email dertroglodyt@gmail.com
 * Copyright by HDC, Germany
 */

class DetailTouchHelper extends ItemTouchHelper.SimpleCallback {

    DetailTouchHelper(ItemDetailAdapter adapter) {
        super(ItemTouchHelper.UP | ItemTouchHelper.DOWN, ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT);
        this.adapter = adapter;
    }

    @Override
    public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
        adapter.swap(viewHolder.getAdapterPosition(), target.getAdapterPosition());
        return true;
    }

    @Override
    public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
        //Remove item
        adapter.remove(viewHolder.getAdapterPosition());
    }

    @Override
    public void onSelectedChanged(RecyclerView.ViewHolder viewHolder, int actionState) {
        if (actionState == ItemTouchHelper.ACTION_STATE_DRAG) {
            adapter.setBGColor(((ItemDetailAdapter.ViewHolder) viewHolder).cb, true);
        }
        super.onSelectedChanged(viewHolder, actionState);
    }

    @Override
    public void clearView(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
        adapter.setBGColor(((ItemDetailAdapter.ViewHolder) viewHolder).cb, false);
        super.clearView(recyclerView, viewHolder);
    }

    private final ItemDetailAdapter adapter;
}

File: app/src/main/java/de/hdc/kspchecklist/DetailTouchHelper.kt
package de.hdc.kspchecklist

import androidx.recyclerview.widget.*

/**
 * Created by DerTroglodyt on 2017-02-18 10:12.
 * Email dertroglodyt@gmail.com
 * Copyright by HDC, Germany
 */

internal class DetailTouchHelper(private val adapter: ItemDetailAdapter)
    : ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP or ItemTouchHelper.DOWN, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {

    override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
        adapter.swap(viewHolder.adapterPosition, target.adapterPosition)
        return true
    }

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
        //Remove item
        adapter.remove(viewHolder.adapterPosition)
    }

    override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
        if (actionState == ItemTouchHelper.ACTION_STATE_DRAG) {
            adapter.setBGColor((viewHolder as ItemDetailAdapter.ViewHolder).cb, true)
        }
        super.onSelectedChanged(viewHolder, actionState)
    }

    override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
        adapter.setBGColor((viewHolder as ItemDetailAdapter.ViewHolder).cb, false)
        super.clearView(recyclerView, viewHolder)
    }
}
